solidity/test/sendERC721ToCosmos.ts

```diff
@@ -0,0 +1,121 @@
+import chai from "chai";
+import { ethers } from "hardhat";
+import { solidity } from "ethereum-waffle";
+
+import { deployContractsERC721 } from "../test-utils/deployERC721";
+import {
+  examplePowers
+} from "../test-utils/pure";
+import { Gravity, GravityERC721, TestERC721A } from "../typechain";
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+
+
+chai.use(solidity);
+const { expect } = chai;
+
+async function runTest(opts: {
+  wrongERC721Owner?: boolean;
+  ERC721NotInContract?: boolean;
+  ERC721NotExist?: boolean; 
+}) {
+
+  // Prep and deploy contract
+  // ========================
+  const signers = await ethers.getSigners();
+  const gravityId = ethers.utils.formatBytes32String("foo");
+  // This is the power distribution on the Cosmos hub as of 7/14/2020
+  let powers = examplePowers();
+  let validators = signers.slice(0, powers.length);
+  const {
+    gravity,
+    gravityERC721,
+    testERC721,
+    fakeGravity,
+    checkpoint
+  } = await deployContractsERC721(gravityId, validators, powers);
+
+  
+  // Transfer out to Cosmos, locking coins
+  // =====================================
+  if (!opts.wrongERC721Owner){
+    await testERC721.functions.approve(gravityERC721.address, 190);
+  }
+    await firstCall(gravityERC721, testERC721, signers, gravity);
+    let secondERC721 = await getSecondERC721(testERC721, gravityERC721, opts.ERC721NotExist, opts.ERC721NotInContract);
+    if (!opts.ERC721NotExist && !opts.ERC721NotInContract) {
+      await testERC721.functions.approve(gravityERC721.address, secondERC721);
+    }
+    await secondCall(gravityERC721, testERC721, signers, gravity, secondERC721);
+}
+
+async function getSecondERC721(testERC721: TestERC721A, gravityERC721: GravityERC721,
```

Comment:
function comment on this would be appreciated, I'm having trouble following what it's doing.

Reply:
Comment added